common/src/main/java/com/wynntils/features/FixPacketBugsFeature.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features;
+
+import com.wynntils.core.features.FeatureBase;
+import com.wynntils.mc.mixin.accessors.ClientboundSetPlayerTeamPacketAccessor;
+import com.wynntils.mc.utils.McUtils;
+import java.util.Map;
+import java.util.UUID;
+import net.minecraft.client.gui.components.LerpingBossEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.game.ClientboundBossEventPacket;
+import net.minecraft.network.protocol.game.ClientboundBossEventPacket.Handler;
+import net.minecraft.network.protocol.game.ClientboundSetPlayerTeamPacket;
+import net.minecraft.world.BossEvent.BossBarColor;
+import net.minecraft.world.BossEvent.BossBarOverlay;
+import net.minecraft.world.scores.PlayerTeam;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+public class FixPacketBugsFeature extends FeatureBase {
+
+    public FixPacketBugsFeature() {
+        setupEventListener();
```

Comment:
Yes, that is one possible design decision. I used to share that opinion, now I'm not as sure. My original thinking was that we should limit our mixins like this:

1) The actual mixin code should only be a single line, calling a method in a non-mixin class. The reasons for this is basically debuggability: mixin classes can't have breakpoints, and can't be hotswapped.

2) This function should in principle create an event and dispatch it, and not do anything else.

3) All other functionality interacting with mixins should do so only via these events.

I still see much value in upholding these limitations. It makes the code base more predictable, it's a well-proven method to keep the separation of concern, and make the flow of code unentangled. 

But I'm also starting to worry that this is too limiting, and/or requiring too much work for "boilerplate" code. Instead of just doing the functionality in the mixin (like for the classes in this PR), we would need to create an EventFactory method, an event type and an event listener in a separate class. This design is more flexible, arguably better, but there is no denying that it is also a lot more work.

And if we want to keep doing events for everything, we need to be able to send back information, like a modified chat message. Right now the only backchannel is cancelling an event. Are we all just fine with event fields being non-final and mutable, and that these fields will get picked up after the event has been processed? It kind of goes against my intuition on what an "event" is, but I see no other good way of passing changes from an event handler back to the mixin.

Reply:
In this particular case I'm also a bit unwilling to create events since these are in a sense not "real" functionality -- it's patchy workarounds to fix annoying bugs -- but it's not bugs that are visible to ordinary users, but mostly just developers.

Otoh, this makes FixPacketBugsFeature not function as an ordinary feature, which can be turned on and off. So either these static functions should move to a bugfix class in e.g. `util`, or it should be turned into a proper Feature, with Events. I'm on the fence here.

Reply:
IMO it's fine not to completely restrict ourselves to events when working with mixins, especially for edge cases like this

regarding whether events should have mutable fields - it's how forge events work already (and we're using their event superclass, after all), so i think there's plenty of precedent for it to make sense

Reply:
I actually managed to persuade myself into keeping with the original design philosophy. :-)

If this is not a proper feature, it will be just weird. With the static callbacks, it can't be disabled, etc. That's just messy.

Also, after thinking about it a bit more, I think it's good to try to actually keep with the "mixins just send events" philosophy. After all, that was how Forge/Legacy Wynntils worked, and it served us just fine. Even if it's a bit more work, I think it will pay off in the long run due to the lessened cognitive load -- we can be sure of that all mixins send events, and that there are no exceptions.

Reply:
i mean, legacy wynntils worked that way because mixins weren't a thing in the first place - events were all we had. i'm not steadfastly opposed to keeping everything in the event framework, but i don't think that limiting ourselves because of what the limits used to be is a particularly strong justification. i'd like to hear @kristofbolyai weigh in on this

Reply:
Like we discussed in the LR PR, I don't think that making everything a feature is necessary good or helps long term, and I don't think we should limit ourselves to events either. 

I don't think this is something that should be a feature, more like a `functionality` we discussed earlier (https://github.com/Wynntils/Artemis/pull/50#discussion_r872972307).

> Once again, I feel we're missing a place to put "Wynntils functionality", that is not a direct part of a feature. Could be have a functionality package, as a peer to features, for Wynntils functionality that is not bundled as a feature to be turned on and off? Otherwise I think the best solution still is to shoehorn the lootruns into a Feature.

Although we can accept this PR, this is the second time in a few PRs that the need for a `functionality` package came up, we should definitely work on that. We will sure see the need for it elsewhere later.

As for original question, no I don't think we should only use events with mixins. We will likely get to a point where we would not be able to achieve a specific feature, just because we limited ourselves to mixins only injecting events. In my opinion, let's use the power of mixins to our advantage, and not limit ourselves. In the end, I agree that we should only call a single method in mixins, just to make then easier to debug and work on, but that single call should not be limited to posting an event.

Reply:
I think we're all in agreement here. Let me summarize:

1) We should strive for having mixins send events, if it makes sense, but if not, we are open to do other things in mixins.

2) Mixins should, if possible, be restricted to a single method call.

3) We really do need some way to handle functionality that is not tied to features. Exactly how this should be designed is still a bit unclear, since we don't know the full scope of responsibilities for that part of the code.

Reply:
> 3\. We really do need some way to handle functionality that is not tied to features. Exactly how this should be designed is still a bit unclear, since we don't know the full scope of responsibilities for that part of the code.

Maybe we could start by making `functionality` similar to features, but:

1.  They can't be turned on and off by the user (or even with code?).
2.  They are not instance based, they only have static methods for events
3.  They are not directly linked to features nor classes, but both of them can have functionality classes to handle `util` like work.
4.  Although not being able to turn them off, they should be also "configurable". This means they could have fields that are marked as configurable and could be changed by user configs. (Although this is how I imagine features as well). This way we can still group settings by their related functions in sections, like we have in legacy right now. (a nice example of functionality are lootrun paths, maps, cloud config saving). 

(We should really start working on configs :^) )

Reply:
Yes, that sounds good. We can transform this feature to a "functionality" as a start. But I'd prefer to push this now.

And yes. Configs. We need them yesterday. :-)

Reply:
Yes, functionality is a separate PR.

Reply:
I checked some synonyms for "functionality" to get inspiration. Perhaps "services" or "components"? (Or we could just start with "functionality" and rename it later if/when we can agree on a better name.)

Reply:
Let's go with "functionality" for now. I believe in moving fast and correcting mistakes later on, rather than dwell too much up front.

Reply:
@magicus are you planning on working on it in the near future? If not, I may as well start working on it.

Reply:
> I think we're all in agreement here. Let me summarize:
> 
> 1. We should strive for having mixins send events, if it makes sense, but if not, we are open to do other things in mixins.
> 2. Mixins should, if possible, be restricted to a single method call.
> 3. We really do need some way to handle functionality that is not tied to features. Exactly how this should be designed is still a bit unclear, since we don't know the full scope of responsibilities for that part of the code.

@Incompleteusern @P0keDev Do you agree? If so, let's merge.